app/services/deal_analytic_service.rb

```diff
@@ -424,6 +429,140 @@ def check_deal_analytics(merchant_id)
       deal_analytics
     end
 
+    # returns conversion rate of num wishlisters to viewers in percentage
+    def get_wishlist_to_views(deal_id)
+      push_date = Deal.find(deal_id).push_date
+      num_wishlist = WishService.num_wishlist_deal(deal_id, push_date).to_f
+      user_id = WishService.get_user_id(deal_id, push_date)
+      num_views = Viewcount.where(user_id: user_id, deal_id: deal_id).uniq.where(:entry => 'merchant_push_notification').count.to_f
+      conversion = (num_views/num_wishlist)*100
+      conversion.round(2)
+    end
+
+    # returns conversion rate of num unique views to redeems in percentage
+    def get_views_to_redeem(deal_id)
+      view_counts = ViewcountService.get_uniq_view_count(deal_id).to_f
+      user_ids = ViewcountService.get_uniq_user_id(deal_id)
+      redeem_count = RedemptionService.count_uniq_redemptions(deal_id, user_ids).to_f
+      conversion = (redeem_count/view_counts)*100
+      conversion.round(2)
+    end
+
+    # returns percentage of user who redeemed more than once
+    def get_multiple_redeems_percentage(deal_id)
+      user_count = RedemptionService.count_uniq_redemptions(deal_id).to_f
+      multiple_redeems = RedemptionService.num_users_multiple(deal_id).to_f
+      percentage = (multiple_redeems/user_count)*100
+      percentage.round(2)
+    end
+
+    # returns average time between users who redeem more than once
+    def average_time_btw_multiple_redeem(deal_id)
+      multiple_redeems_user_ids = RedemptionService.get_user_ids(deal_id, true)
+      if multiple_redeems_user_ids.blank?
+        'No Redeems Yet'
+      else
+        (RedemptionService.get_average_days_between_redeems(deal_id, multiple_redeems_user_ids).to_s + ' days')
+      end
+    end
+
+    # returns a nested array
+    # array[0] is cumulative increase in percentage per hour
+    # array[1] is change in percentage per hour
+    def get_wishlist_to_view(deal_id)
+      array = Array.new
+      cumulative = Array.new
+      non_cumulative = Array.new
+      push_date = Deal.find(deal_id).push_date.beginning_of_hour.to_datetime
```

Comment:
@Woonyong92 Will return error if push_date unset - should disable "Show Charts" button if deal has not been pushed or do additional checks to avoid error.